File: leetcode/src/main/java/leetcode/FirstMissingPositive.java
package leetcode;

public class FirstMissingPositive {
  public int firstMissingPositive(int[] nums) {
    for (int i = 0; i < nums.length; i++) {
      while (0 < nums[i] && nums[i] <= nums.length && nums[i] != nums[nums[i] - 1]) {
        swap(i, nums[i] - 1, nums);
      }
    }

    for (int i = 0; i < nums.length; i++) {
      if (nums[i] != i + 1) {
        return i + 1;
      }
    }
    return nums.length + 1;
  }

  private void swap(int i, int j, int[] nums) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/FirstMissingPositive.kt
package com.giwankim.leetcode

class FirstMissingPositive {
    fun firstMissingPositive(nums: IntArray): Int {
        for (i in nums.indices) {
            while (nums[i] in 1..nums.size && nums[i] != nums[nums[i] - 1]) {
                nums.swap(i, nums[i] - 1)
            }
        }
        for (i in nums.indices) {
            if (nums[i] != i + 1) {
                return i + 1
            }
        }
        return nums.size + 1
    }

    fun IntArray.swap(
        i: Int,
        j: Int,
    ) {
        val temp = this[i]
        this[i] = this[j]
        this[j] = temp
    }
}
